python/ray/_private/worker.py

```diff
@@ -125,7 +125,40 @@
 DAGNode = TypeVar("DAGNode")
 
 
-class RemoteFunctionNoArgs(Generic[R]):
+# Only used for type annotations as a placeholder
+Undefined: Any = object()
+
+
+# TypeVar for self-referential generics in `RemoteFunction[N]`.
+RF = TypeVar("RF", bound="HasOptions")
+
+
+class HasOptions(Protocol):
+    def options(
+        self: RF,
+        num_returns: Union[int, Literal["streaming"]] = Undefined,
+        num_cpus: Union[int, float] = Undefined,
+        num_gpus: Union[int, float] = Undefined,
+        resources: Dict[str, float] = Undefined,
+        accelerator_type: str = Undefined,
+        memory: Union[int, float] = Undefined,
+        max_calls: int = Undefined,
+        max_restarts: int = Undefined,
+        max_task_retries: int = Undefined,
+        max_retries: int = Undefined,
+        runtime_env: Dict[str, Any] = Undefined,
+        retry_exceptions: bool = Undefined,
+        scheduling_strategy: Union[
+            None,
+            Literal["DEFAULT"],
+            Literal["SPREAD"],
+            PlacementGroupSchedulingStrategy,
+        ] = Undefined,
```

Comment:
Question: how do we keep these in sync with the actual args? In other words, can/should we just use `def options(self, **task_options)` ?

Reply:
That is an improvement, thank you

Reply:
Confirms that my reproduction script above still passes typecheck with this change